docs/contribution-guides/Development-Standards-Guide.md

```diff
@@ -226,3 +242,301 @@ to just leave it as a public field without a property. As always, make sure to d
 6. Most string or numeric literals should be constants.
 7. Don't define constants that have the same value in multiple places. There should only ever be one place you need to change if you ever need to change a constant's value.
         
+# Writing efficent and maintainble code
+
+Unitystation is a gigantic and ambitious project that holds a lot of content and systems that can get tangled together, to avoid things tangling each other; we highly recommend that you write your code in a modular way that's easy for developers to maintain and and expand upon.
+
+Unity has a great talk on how to do just that in this video: [How to adopt a modular code workflow for your team | Unite 2022](https://youtu.be/nk3gHIZZ5Rg)
+
+Below are good practices and explainations on what to do, so you can avoid having your code refactored/reworked 5 months later and make the lives of other contributors much easier.
+
+## 0. In-code documentation
+
+* You can avoid confusing others (and even yourself) by naming your functions and variables in a clear and descriptive way.
+
+```cs
+//bad code
+void doThing(bool b)
+{
+    state = b;
+}
+```
+
+```cs
+//good code
+private void ThisFunctionWillDoA_B_And_C(bool newState)
+{
+    state = newState;
+}
+```
+
+* Explain what your function does step by step whenever possible. Your code might seem obvious to you, but for other people; that may not be the case.
```

Comment:
Jesus no, we don't need a comment every step. In the summary it is fine, but only required for public methods. A private method could have a summary if the thing it does is particularly obscure

Reply:
I will try to be precise and concise:

### I . **Too many comments in code is actually a detriment to understanding the code**.

If you document literally every step, like you are asking here, I will expend more time reading comments about how something is supposed to work instead of how it is actually working. I could ignore the logic presented and believe the comment, when that might not actually be the truth. And I should also mention the clutter.

### II . **The need for a lot of comments is actually hiding bad code**.

If you feel the need to document every step, it could be that the algorithm you have written is simply hard to understand and isn't using proper names for variables/methods. When code is well written, you can understand it by simply reading it. Naming methods is a powerful way to let people know what the algorithm is supposed to do. As a general rule, you should let the code itself say the ``how`` and if it is really necessary, you can add a ``why`` in comments.

### III . **Unitystation as a learning experience**.

Unitystation is indeed a powerful learning experience, but it shouldn't be the learning experience for someone who doesn't know programming at all. We shouldn't add comments like
```cs
// returns early if something is null
if (something == null) return;
```
If that explanation was needed, then the person should first try in smaller projects of their own to get a grasp on programming before trying to contribute to Unitysation.

Also take in cosideration the **source code** is for developers, people who aren't writing code shouldn't need to read it. We have other instances for those cases.

### IV . **If extensive explanation of a system is needed, the source code isn't the place for that explanation**

In some cases, the systems we are building can be complex enough that a thorough explanation is required. You won't understand the system by reading literally every comment we wrote line by line of code in the system. We have a proper development wiki for that pupose where a well thought documentation can be written so the reader gets a birdeye view of the system and then a dense explanation of all its parts, including examples, etc.

### V . **Literally no one ever has ever needed commenting every step. We would be the first codebase to do that**.

And I don't mind being pioneer, but in this case it is simply because it is a wrong move. If you lurk around, you will find that most people when talking about "too many comments" are referring to summaries of algorithms, commenting literally every step would be even worse than "too many comments".

I will leave you some random articles I found on google so you can get a better written explanation on the concept:

[Portrait of a n00b](https://steve-yegge.blogspot.com/2008/02/portrait-of-n00b.html)
[Fighting Evil in Your Code: Comments on Comments](https://www.red-gate.com/simple-talk/opinion/opinion-pieces/fighting-evil-code-comments-comments/)
[You’re commenting your code too much (and other controversial thoughts on documentation)](https://levelup.gitconnected.com/youre-commenting-your-code-too-much-and-other-controversial-thoughts-on-documentation-1ee617ed46af)

Reply:
> I . Too many comments in code is actually a detriment to understanding the code.

I can make a case and say that we don't need to do it all the time, but we still should enforce it in a lot of cases.

> II . The need for a lot of comments is actually hiding bad code.

Yeah, no shit? 

This has been a thing that I've been complaining about earlier than #8529 and has gotten worse with #9178 when our lead developer literally **threatened** to veto contributions that aim to **improve** bad code because he's stubborn and does not want anyone to challenge things he poorly implemented, and he still continues to push code that is difficult to understand and work with until as recently as effect syncs with #9321 .

Since no one wanted to make an effort to improve the code quality recently after all my begging to do so and I still have to re-write scripts to improve its general use and understanding; then at least explain to me what the hell your bad code does in the first place.

I don't understand why this is such a difficult thing to ask, either treat Bod's PRs with the same level of annoying semi-strictness as me when it comes to my PRs or make it mandatory for him to **at least** make the effort into explaining what his bad code does.

> Unitystation as a learning experience.

Like I've stated above, I'm willing to reduce how mandatory it is to document steps that don't require docs to start with; but we have to still tell people to explain what their code does when it starts to get technical or large.

a good example:
```cs
/// <summary>
/// Checks if the item is in the correct ItemSlot which is the eyes.
/// Automatically returns false if null because of the "is" keyword and null propagation.
/// </summary>
private bool IsInCorrectNamedSlot()
{
	return pickupable.ItemSlot is { NamedSlot: NamedSlot.eyes };
}
```

> IV . If extensive explanation of a system is needed, the source code isn't the place for that explanation

In any other circumstance I would have agreed with this system because even I complained to SS14 devs about them not having any sort of extensive developer docs like [we do](https://unitystation.github.io/unitystation/development/Helpful-Functions-Cheatsheet/) 

But all I have to point to is #8830 to show you how this is going to be a failed thing to push.

"So, you want us to do [bad thing] instead??"

Yes, since there is no viable alternative; I'd rather that we right a wrong with another wrong.

> Literally no one ever has ever needed commenting every step. We would be the first codebase to do that.

I agree with this, but I still want to push this standard in some form of way or another because this is important to the codebase's quality in the future since we shouldn't have only one person who understands how everything works in this project and does not want to cooperate with you fully to add even the most basic of features sometimes.

If you want to discuss the best middle ground we can have, we can discuss that on discord.

Reply:
Done, review that again please.